posthog/models/async_deletion/delete.py

```diff
@@ -0,0 +1,143 @@
+from collections import defaultdict
+from typing import Any, Dict, List, Set, Tuple
+
+import structlog
+from django.utils import timezone
+
+from posthog.client import sync_execute
+from posthog.models.async_deletion import AsyncDeletion, DeletionType
+from posthog.settings.data_stores import CLICKHOUSE_CLUSTER
+
+logger = structlog.get_logger(__name__)
+# Note: Session recording, dead letter queue, logs deletion will be handled by TTL
+TABLES_TO_DELETE_TEAMS_FROM = [
+    "person",
+    "person_distinct_id",
+    "person_distinct_id2",
+    "groups",
+    "cohortpeople",
+    "person_static_cohort",
+    "plugin_log_entries",
+]
+
+
+def run_event_table_deletions():
+    deletions = list(AsyncDeletion.objects.filter(delete_verified_at__isnull=True))
+
+    if len(deletions) == 0:
+        logger.debug("No AsyncDeletion to perform")
+        return
+
+    logger.info(
+        "Starting AsyncDeletion on `events` table in ClickHouse",
+        {"count": len(deletions), "team_ids": list(set(row.team_id for row in deletions)),},
+    )
+
+    conditions, args = _conditions(deletions)
+    sync_execute(
+        f"""
+        ALTER TABLE sharded_events
+        ON CLUSTER '{CLICKHOUSE_CLUSTER}'
+        DELETE WHERE {" OR ".join(conditions)}
+        """,
+        args,
+    )
+
+    # Team data needs to be deleted from other models as well, groups/persons handles deletions on a schema level
+    team_deletions = [row for row in deletions if row.deletion_type == DeletionType.Team]
+
+    if len(team_deletions) == 0:
+        return
+
+    logger.info(
+        "Starting AsyncDeletion for teams on other tables",
+        {"count": len(team_deletions), "team_ids": list(set(row.team_id for row in deletions)),},
+    )
+    conditions, args = _conditions(team_deletions)
+    for table in TABLES_TO_DELETE_TEAMS_FROM:
+        sync_execute(
+            f"""
+            ALTER TABLE {table}
+            ON CLUSTER '{CLICKHOUSE_CLUSTER}'
+            DELETE WHERE {" OR ".join(conditions)}
```

Comment:
q: say this query gets massively large - is that a problem for CH to handle? I assume no but just checking

Reply:
Define massively large and problem to handle?

We're bottlenecked currently on rewriting $ALL_OF_DATA on disk and I don't see ~20 conditions suddenly making this CPU-bound. If there's query complexity issues we have Sentry for that.